### PR TITLE
OpenML API

### PR DESCRIPTION
**Pull request by darenasc**
OpenML.org API integration, the `OpenML.load()` function will download a dataset from the OpenML platform given its ID number, and return it as a `Vector` of `NamedTuples`.

For example, the iris dataset has the ID 61 in OpenML. The named tuples can be converted to tabular format such as a `DataFrame`.

```
iris_named_tuples = OpenML.load(61)
df = DataFrame(iris_named_tuples)
```

### CODE REVIEW

File: src/openml.jl [hunk 1]
```diff
@@ -0,0 +1,230 @@
+module OpenML
+
+using HTTP
+using JSON
+
+api_url = "https://www.openml.org/api/v1/json"
```
Comment (tlienart) [inline src/openml.jl]: change for

```julia
const API_URL ...
```

File: src/openml.jl [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+Returns information about a dataset. The information includes the name,
+information about the creator, URL to download it and more.
+
+110 - Please provide data_id.
+111 - Unknown dataset. Data set description with data_id was not found in the database.
+112 - No access granted. This dataset is not shared with you.
+"""
+function load_Dataset_Description(id::Int; api_key::String="")
+    url = string(api_url, "/data/$id")
+    try
```
Comment (tlienart) [inline src/openml.jl]: Here you could add a request to Julia's discourse, if it fails you can set a flag to OFFLINE and show an  error message that the user seems offline.

File: src/openml.jl [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+            println("Please provide data_id.")
+        elseif r.status == 111
+            println("Unknown dataset. Data set description with data_id was not found in the database.")
+        elseif r.status == 112
+            println("No access granted. This dataset is not shared with you.")
+        end
+    catch e
+        return "Error occurred : $e"
+    end
+end
```
Comment (tlienart) [inline src/openml.jl]: the end should look like:

```julia
catch e
  @warn "Error occurred : $e"
  return nothing
end
return nothing # for all the  cases where you do  `println`
```

also replace the  `println` with `@warn`

File: src/openml.jl [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+function convert_ARFF_to_rowtable(response)
+    data = String(response.body)
+    data2 = split(data, "\n")
+
+    featureNames = String[]
+    dataTypes = String[]
+    named_tuples = []
+    for line in data2
+        if length(line) > 0
+            if line[1:1] != "%"
```
Comment (tlienart) [inline src/openml.jl]: this is weird, `line[1] != '%'` is better (note single quote)

File: src/openml.jl [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+                    push!(featureNames, replace(split(line, " ")[2], "'" => ""))
+                    push!(dataTypes, split(line, " ")[3])
+                elseif occursin("@relation", lowercase(line))
+                    nothing
+                elseif occursin("@data", lowercase(line))
+                    # it means the data starts so we can create the data frame
+                    nothing
+                else
+                    values = split(line, ",")
+                    for i = 1:length(featureNames)
```
Comment (tlienart) [inline src/openml.jl]: can use `for  i in eachindex(featureNames)` but doesn't matter

File: src/openml.jl [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+                        if in(lowercase(dataTypes[i]), ["real","numeric"])
```
Comment (tlienart) [inline src/openml.jl]: prefer the operation syntax `lowercase(...) in [...]` as it's more readable

File: src/openml.jl [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+370 - No data qualities available. There are no data qualities in the system.
+"""
+function load_Data_Qualities_List()
+    url = string(api_url, "/data/qualities/list")
+    try
+        r = HTTP.request("GET", url)
+        if r.status == 200
+            return JSON.parse(String(r.body))
+        elseif r.status == 370
+            println("No data qualities available. There are no data qualities in the system.")
```
Comment (tlienart) [inline src/openml.jl]: again `@warn` and return something after the try catch

File: src/openml.jl [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    elseif r.status == 271
+        println("Unknown dataset. Data set with the given data ID was not found (or is not shared with you).")
+    elseif r.status == 272
+        println("No features found. The dataset did not contain any features, or we could not extract them.")
+    elseif r.status == 273
+        println("Dataset not processed yet. The dataset was not processed yet, features are not yet available. Please wait for a few minutes.")
+    elseif r.status == 274
+        println("Dataset processed with error. The feature extractor has run into an error while processing the dataset. Please check whether it is a valid supported file. If so, please contact the API admins.")
+    end
+
```
Comment (tlienart) [inline src/openml.jl]: same here

File: src/openml.jl [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+        println("Unknown dataset. The data set with the given ID was not found in the database, or is not shared with you.")
+    elseif r.status == 362
+        println("No qualities found. The registered dataset did not contain any calculated qualities.")
+    elseif r.status == 363
+        println("Dataset not processed yet. The dataset was not processed yet, no qualities are available. Please wait for a few minutes.")
+    elseif r.status == 364
+        println("Dataset processed with error. The quality calculator has run into an error while processing the dataset. Please check whether it is a valid supported file. If so, contact the support team.")
+    elseif r.status == 365
+        println("Interval start or end illegal. There was a problem with the interval start or end.")
+    end
```
Comment (tlienart) [inline src/openml.jl]: and  here

File: src/openml.jl [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+        return JSON.parse(String(r.body))
+    elseif r.status == 370
+        println("Illegal filter specified.")
+    elseif r.status == 371
+        println("Filter values/ranges not properly specified.")
+    elseif r.status == 372
+        println("No results. There where no matches for the given constraints.")
+    elseif r.status == 373
+        println("Can not specify an offset without a limit.")
+    end
```
Comment (tlienart) [inline src/openml.jl]: and here

File: test/openml.jl [hunk 11]
```diff
@@ -0,0 +1,28 @@
+using Test
```
Comment (tlienart) [inline test/openml.jl]: did you add those tests in runtests.jl?

File: src/openml.jl (hunk 3 above)
Comment (darenasc) [reply src/openml.jl]: done.

File: src/openml.jl (hunk 5 above)
Comment (darenasc) [reply src/openml.jl]: modified.

File: src/openml.jl (hunk 6 above)
Comment (darenasc) [reply src/openml.jl]: yeah, looks more readable, changed.

File: test/openml.jl (hunk 11 above)
Comment (darenasc) [reply test/openml.jl]: thanks, it is added now.

File: src/openml.jl (hunk 4 above)
Comment (darenasc) [reply src/openml.jl]: sorry, it didn't work, I kept it with double quote.

**Comment by ablaom** [comment]
@tlienart Thanks for reviewing!

**Comment by ablaom** [comment]
@darenasc Looks like there's just the one suggestion of @tlienart about testing the network is online. Did you want to do something about this?

**Comment by darenasc** [comment]
Thanks @ablaom, I'll leave that one for an update @tlienart, we don't want to generate a DoS of the discourse page from MLJ :) 

There are error messages are printed for all the statuses and that should work for now.

File: src/openml.jl (hunk 2 above)
Comment (darenasc) [reply src/openml.jl]: Thanks @tlienart I think something around that would be good, maybe using a dedicated service for that or even an internal feature in the HTTP library that can check it. For this PR I think it is ready to go, and I'll see what would be a good solution for that checking and include it in an update soon. There are more functions using the API key that I'd like to work next.

**Pull request merged**